Update API version from v0.3 to v0.4

radumarg opened the pull request:
Update API version from v0.3 to v0.4

**Related Shortcut Stories:**
[sc-115352]

radumarg committed:
Started on updating the API.

radumarg committed:
Update circuit type.

radumarg committed:
More work.

radumarg committed:
More work on prodcution code. Update and extend tests.

radumarg committed:
Merge branch 'master' into new_api_v_04

radumarg committed:
Cosmetic.

radumarg committed:
Update changlog. Improve tests. Handle dry run.

radumarg committed:
Adding missing job settings(params) + tests.

radumarg committed:
Last fixes. Ready for review.

radumarg committed:
Remove print statements,

radumarg committed:
Improve docstrings.

splch committed:
Refactor BASE_URL initialization and clean up job creation documentation

splch committed:
Fix conditional check for job settings initialization in IonQDevice

splch committed:
Clarify documentation for 'debiasing' parameter in IonQDevice

splch committed:
Simplify result probability URL extraction and API call in ResourceManager

splch committed:
Merge branch 'master' into new_api_v_04

radumarg committed:
Rework after review.

splch committed:
Merge branch 'master' into new_api_v_04

Alex-Preciado commented:
Hey @radumarg @splch, thanks for opening this PR! We were curious if you're still looking to merge this at some point, or if it was mainly an exploratory PR to see what was required for the v0.3 to v0.4 API update? We're happy to support with a code review if you’re still hoping to get this merged!

splch committed:
Fix bugs, simplify, and align with IonQ v0.4 API spec

- Fix gate["time"] double-assignment that overwrote abs()/float() conversion
- Fix shots default to None (consistent with PR #149 on master)
- Omit shots from payload when None (API rejects null, defaults to 100)
- Fix sharpen default to None in IonQDevice (was False, causing unnecessary param)
- Merge _extract_evolution_coefficients and _extract_evolution_pauli_terms into
  single _decompose_evolution method, eliminating duplicate pauli_decompose calls
- Add None guard for coefficients (was crashing with TypeError)
- Extract _append_gate helper to deduplicate circuit-append logic
- Move dry_run guard before polling loop in _submit_job (skip unnecessary I/O)
- Fix CHANGELOG: move v0.4 entry under Breaking Changes, deduplicate contributors
- Fix QPUDevice docstring URL to point to v0.4 API docs
- Fix compilation docstring to proper RST format
- Add missing (bool) type to dry_run docstring
- Remove unused imports (XX, YY, ZZ in tests; typing.List in device)
- Fix _remove_trivial_terms to use idiomatic zip and not-in syntax
- Fix test assertions (== True -> is True), grammar, add missing shots param
- Fix typo in exceptions.py (coeffcient -> coefficient)

splch committed:
Merge branch 'master' into new_api_v_04

splch committed:
Merge upstream/master, resolve conflicts, fix Ising gate tests

- Keep upstream sign fix (np.sign * -1) for Evolution gate coefficients
- Adapt upstream Ising gate tests to single-circuit input format
- Restore XX/YY/ZZ imports needed by new upstream tests

splch committed:
Fix dry_run multi-circuit return, clean up reset and histogram logic

- Fix batch_execute dry_run to return one result per circuit, not [[]]
- Use is not None for noise/metadata checks (consistent with other params)
- Cache generator.base.terms() result instead of calling twice
- Simplify histogram construction with list(values())
- Deduplicate input dict construction in reset()

splch committed:
Fix black formatting, remove unused loop variable, clean up

- Run black formatter on all source and test files (fixes CI black check)
- Remove unused loop variable i in apply()
- Simplify histogram construction

splch committed:
Fix black formatting to match CI config (line-length 100)

splch commented:
> Hey @radumarg @splch, thanks for opening this PR! We were curious if you're still looking to merge this at some point, or if it was mainly an exploratory PR to see what was required for the v0.3 to v0.4 API update? We're happy to support with a code review if you’re still hoping to get this merged!

Hi @Alex-Preciado! Thank you for checking in. We definitely want to get this merged. We've done enough clean ups now where we'd welcome a code review whenever someone has time :)

The codecov failures are a fork PR limitation so I think everything else is looking good. Once we have your approval, we're happy to merge on our end.

JerryChen97 commented on CHANGELOG.md:
```diff
@@ -70,8 +74,9 @@ Andrija Paurevic.
+Andrija Paurevic,
 Radu Marginean,
-Andrija Paurevic.
+Spencer Churchill.
```
Mind changing it to alphabetical order? I was about to write a suggestion but GH doesn't allow me to make suggestions on deleted lines somehow

JerryChen97 commented on CHANGELOG.md:
Let's use the changelog region for 0.45.0-dev!

JerryChen97 commented on pennylane_ionq/device.py:
```diff
@@ -100,20 +99,28 @@ class IonQDevice(QubitDevice):
-            Valid keys include: ``debias`` (bool).
-        sharpen (bool): whether to use sharpening when accessing the results of an executed job. Defaults to None
+            Valid keys include: ``debiasing`` (bool).
+        sharpen (bool): Whether to use sharpening when accessing the results of an executed job. Defaults to None
```
```suggestion
        sharpen (bool): whether to use sharpening when accessing the results of an executed job. Defaults to None

```

JerryChen97 commented on pennylane_ionq/exceptions.py:
Thanks!!

splch committed:
lowercase whether

Co-authored-by: Yushao Chen (Jerry) <chenys13@outlook.com>

splch committed:
Move PR #140 changelog entry to 0.45.0-dev, fix contributor order

Address review feedback from JerryChen97:
- Move v0.4 API entry from released 0.43.0 section to 0.45.0-dev
- Alphabetize contributors in both sections
- Remove PR-specific contributors from 0.43.0 section

splch commented:
Thank you for the reviews @JerryChen97! I got the fixes in :)

JerryChen97 commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -130,6 +130,48 @@ def test_evolution_gate_print_warning(self, requires_api):
+        H = qml.X(0)
+        tape = qml.tape.QuantumScript([qml.evolve(H, time)], [qml.probs(wires=[0])])
+
+        result_ionq = dev.batch_execute([tape])
```
```suggestion
        dev = qml.device("ionq.simulator", wires=2, gateset="qis")

        time = 1
        H = qml.X(0)
        tape = qml.tape.QuantumScript([qml.evolve(H, time)], [qml.probs(wires=[0])], shots=1024)

        result_ionq = dev.batch_execute([tape])
```

JerryChen97 commented on tests/test_pauliexp_ionq_gate.py:
```diff
@@ -130,6 +130,48 @@ def test_evolution_gate_print_warning(self, requires_api):
+        H1 = qml.X(0)
+        H2 = qml.Z(0)
+        tape1 = qml.tape.QuantumScript([qml.evolve(H1, time)], [qml.probs(wires=[0])])
+        tape2 = qml.tape.QuantumScript([qml.evolve(H2, time)], [qml.probs(wires=[0])])
```
```suggestion
        dev = qml.device("ionq.simulator", wires=2, gateset="qis")

        time = 1
        H1 = qml.X(0)
        H2 = qml.Z(0)
        tape1 = qml.tape.QuantumScript([qml.evolve(H1, time)], [qml.probs(wires=[0])], shots=1024)
        tape2 = qml.tape.QuantumScript([qml.evolve(H2, time)], [qml.probs(wires=[0])], shots=1024)
```

JerryChen97 reviewed (changes requested):
Thanks a lot for shipping the API update! I think I understand the new API now. Here are some things that are actually on our side that happened recently:
1. we deprecated the `device(..., shots=...)` api. Although this interface is still available, whenever we explicitly use it there will be a `PennyLaneDeprecationWarning` emitted. In the past when I deprecated it in ionq plugin, I played some trick such that ionq device can just use the shots information from tapes directly. I left some suggestions in the tests so that this was also covered instead of the deprecated interface, and hopefully the refactoring happens in `IonQDevice` and `SimulationDevice` is working well!
2. I believe the integration test CI was not triggered since our new security policy blocked any trigger from external users. I'll branch out and upload a new PR to trigger this. This probably is also the reason why CodeCov complaint a lot. Sorry for the inconvenience!

splch committed:
Use tape-level shots instead of deprecated device-level shots

Address JerryChen97 review: use QuantumScript(..., shots=) instead of
qml.device(..., shots=) in integration tests since device-level shots
are deprecated in PennyLane >= 0.44.

batch_execute now picks up shots from tapes when device shots are None,
both for the API payload and for generate_samples().

splch commented:
Thanks for the thorough reviews @JerryChen97! All feedback has been addressed:
- Moved the #140 changelog entry to the 0.45.0-dev section with contributors in alphabetical order
- Updated the two integration tests to use tape-level shots= via QuantumScript(..., shots=1024) instead of the deprecated qml.device(..., shots=)
- batch_execute now picks up shots from tapes when device shots are None, so the new tape-level shots flow works end-to-end (both for the API payload and generate_samples())

Good to know about the codecov upload issue on fork PRs that explains the inflated missing-line count. Locally we see 98% coverage on device.py.

Please let us know if there's anything else :)

andrijapau commented:
> Thanks for the thorough @JerryChen97! All feedback has been addressed:
> 
>     * Moved the [Update API version from v0.3 to v0.4 #140](https://github.com/PennyLaneAI/PennyLane-IonQ/pull/140) changelog entry to the 0.45.0-dev section with contributors in alphabetical order
> 
>     * Updated the two integration tests to use tape-level shots= via QuantumScript(..., shots=1024) instead of the deprecated qml.device(..., shots=)
> 
>     * batch_execute now picks up shots from tapes when device shots are None, so the new tape-level shots flow works end-to-end (both for the API payload and generate_samples())
> 
> 
> Good to know about the codecov upload issue on fork PRs that explains the inflated missing-line count. Locally we see 98% coverage on device.py.
> 
> Please let us know if there's anything else :)

Thanks @splch , just writing here to let you know that we require two technical approvals before merging into `main` so I'll be helping out here as well! 😃 

I am aiming to complete a thorough review by EOD tomorrow so we can get this PR in this week! 💪🏼

splch commented:
Thanks @andrijapau! We appreciate the quick and helpful feedback :)

JerryChen97 commented on pennylane_ionq/api_client.py:
```diff
@@ -353,9 +353,10 @@ def refresh_data(self, data, params=None):
-        if "results_url" in data.keys():
-            result = self.client.get(self.join_path(data["results_url"]), params=params)
-            self.resource.fields[-1].set(result.json())
+        url = ((data.get("results") or {}).get("probabilities") or {}).get("url")
```
```suggestion
        results = data.get("results") or {}
        probabilities = results.get("probabilities") or {}
        url = probabilities.get("url")
```
I personally prefer readability than compression :face_with_peeking_eye:

andrijapau replied:
I think `url = data.get("results", {}).get("probabilities", {}).get("url")` is better IMO

JerryChen97 commented on tests/test_device.py:
```diff
@@ -127,19 +134,22 @@ def test_shots(self, shots, monkeypatch, mocker, tol):
         monkeypatch.setattr(ResourceManager, "get", fake_response)
 
-        dev = qml.device("ionq.simulator", wires=1, api_key="test")
+        dev = qml.device("ionq.simulator", wires=1, shots=shots, api_key="test")
```
```suggestion
        dev = qml.device("ionq.simulator", wires=1, api_key="test")
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -127,19 +134,22 @@ def test_shots(self, shots, monkeypatch, mocker, tol):
+        dev = qml.device("ionq.simulator", wires=1, shots=shots, api_key="test")
 
-        @qml.set_shots(shots)
         @qml.qnode(dev)
```
```suggestion
        @qml.qnode(dev, shots=shots)
```

JerryChen97 replied:
Or equivalently just revert the change

andrijapau commented on pennylane_ionq/api_client.py:
```diff
@@ -353,9 +353,10 @@ def refresh_data(self, data, params=None):
-        if "results_url" in data.keys():
-            result = self.client.get(self.join_path(data["results_url"]), params=params)
-            self.resource.fields[-1].set(result.json())
+        url = ((data.get("results") or {}).get("probabilities") or {}).get("url")
```
```suggestion
        url = data.get("results", {}).get("probabilities", {}).get("url")
```
A bit easier to read IMO!

andrijapau commented on pennylane_ionq/device.py:
```diff
@@ -100,20 +99,29 @@ class IonQDevice(QubitDevice):
+        shots (int, list[int], None): Number of circuit evaluations/random samples used to estimate
+            expectation values of observables. Defaults to None.
             If a list of integers is passed, the circuit evaluations are batched over the list of shots.
+        job_name (str): Optional job name. Defaults to None.
```
```suggestion
        job_name (str | None): Optional job name. Defaults to None.
```

andrijapau commented on pennylane_ionq/device.py:
```diff
@@ -100,20 +99,29 @@ class IonQDevice(QubitDevice):
             (no value passed at job retrieval). Will generally return more accurate results if your expected output
             distribution has peaks. See `IonQ Debiasing and Sharpening
             <https://ionq.com/resources/debiasing-and-sharpening>`_ for details.
+        noise (dict): {"model": str, "seed": int or None}. Defaults to None.
```
```suggestion
        noise (dict | None): {"model": str, "seed": int or None}. Defaults to None.
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+            "ionq.simulator", wires=1, shots=1024, dry_run=True, api_key="test"
+        )
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device(
            "ionq.simulator", wires=1, dry_run=True, api_key="test"
        )

        @qml.qnode(dev, shots=1024)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+
+        dev = qml.device("ionq.simulator", wires=1, shots=1024, dry_run=True)
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device("ionq.simulator", wires=1, dry_run=True)

        @qml.qnode(dev, shots=1024)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+            api_key="test",
+        )
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device(
            "ionq.simulator",
            wires=1,
            noise={"model": "ideal", "seed": 7},
            api_key="test",
        )

        @qml.qnode(dev, shots=1024)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+            api_key="test",
+        )
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device(
            "ionq.simulator",
            wires=1,
            metadata={"key": "value"},
            api_key="test",
        )

        @qml.qnode(dev, shots=1024)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+            "ionq.simulator", wires=1, shots=1024, job_name="test_name", api_key="test"
+        )
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device(
            "ionq.simulator", wires=1, job_name="test_name", api_key="test"
        )

        @qml.qnode(dev, shots=1024)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
+            compilation=compilation,
+        )
+
+        @qml.qnode(dev)
```
```suggestion
        dev = qml.device(
            "ionq.qpu",
            wires=1,
            api_key="test",
            compilation=compilation,
        )

        @qml.qnode(dev, shots=5000)
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
         dev = qml.device(
             "ionq.qpu",
             wires=1,
+            shots=5000,
```
```suggestion
```

JerryChen97 commented on tests/test_device.py:
```diff
@@ -150,31 +160,247 @@ def circuit():
         )
 
-        @qml.set_shots(5000)
         @qml.qnode(dev)
```
```suggestion
        @qml.qnode(dev, shots=5000)
```

andrijapau commented on pennylane_ionq/device.py:
```diff
@@ -626,20 +676,22 @@ class QPUDevice(IonQDevice):
             your expected output distribution has peaks. See `IonQ Debiasing and Sharpening
             <https://ionq.com/resources/debiasing-and-sharpening>`_ for details.
+        dry_run (bool): whether to run the job in dry run mode. Defaults to False.
+        metadata (dict): optional metadata to attach to the job. Defaults to None.
```
```suggestion
        metadata (dict | None): optional metadata to attach to the job. Defaults to None.
```

andrijapau commented on pennylane_ionq/device.py:
```diff
@@ -100,20 +99,29 @@ class IonQDevice(QubitDevice):
+        job_name (str): Optional job name. Defaults to None.
         api_key (str): The IonQ API key. If not provided, the environment
             variable ``IONQ_API_KEY`` is used.
+        compilation (dict): Settings for compilation when creating a job. Defaults to None.
```
```suggestion
        compilation (dict | None): Settings for compilation when creating a job. Defaults to None.
```

andrijapau commented on CHANGELOG.md:
```diff
@@ -69,9 +75,8 @@ Andrija Paurevic.
 This release contains contributions from (in alphabetical order):
 
-Yushao Chen,
-Radu Marginean,
```
Why did you remove @radumarg here?

andrijapau reviewed (commented):
Just some comments on the source code ⏬

splch committed:
Address review feedback from JerryChen97 and andrijapau

- Break nested .get() chain into 3 lines for readability (Jerry).
  Keep `or {}` instead of `.get(key, {})` because the IonQ API can
  return `"results": null` which `.get("results", {})` does not catch.
- Move shots from qml.device() to @qml.qnode(shots=) in all unit
  tests to avoid deprecated device-level shots API (Jerry).
- Use `dict | None` and `str | None` type syntax in docstrings to
  match PennyLane upstream style (Andrija).
- Restore Radu Marginean to 0.43.0 contributors (Andrija).

splch committed:
lowercase whether

splch committed:
Apply remaining dict | None docstring updates

Fix 4 missed instances: error_mitigation in IonQDevice and QPUDevice,
noise and metadata in SimulatorDevice. Also remove stray blank line
in sharpen docstring.

JerryChen97 commented on tests/test_pauliexp_ionq_gate.py:
I would suggest also apply our current strategy of setting `tape.copy(shots=None)` for the "correct" result that we want to compare with. Otherwise, the experiment that two finite-shot (and with very limited number of shots) simulations compare with each other would suffer from very high chance of unwanted flaky failures, e.g. https://github.com/PennyLaneAI/PennyLane-IonQ/actions/runs/24042610077/job/70117930394?pr=166

splch committed:
Use tape.copy(shots=None) for simulator comparison in evolution tests

Avoids flaky failures from comparing two finite-shot simulations.
Matches the pattern used by all other tests in this file.

splch committed:
Widen tolerance on flaky sign correctness test (1e-2 to 5e-2)

With 1024 shots, the standard error is ~0.031, so atol=0.01 only
passes when noise is below 1/3 of a standard deviation (~25% failure
rate). Widening to 5e-2 makes the test reliable while still catching
sign errors (which produce differences of ~1.7, not ~0.03).

splch commented:
@JerryChen97 The flaky `test_evolution_sign_correctness[-0.5]` test had atol=0.01 with only 1024 shots - statistically guaranteed to fail 25% of the time. I just widened it to atol=0.05, which is still tight enough to catch real sign bugs (a sign error would produce a delta of ~1.7, not 0.03).

JerryChen97 commented:
Finally we have the correct codecov report after having the CI fully pass! Let me scan one last time :pray: thanks for patience these days

JerryChen97 commented on pennylane_ionq/device.py:
```diff
@@ -230,9 +268,15 @@ def batch_execute(self, circuits):
         for circuit_index, circuit in enumerate(circuits):
             self.set_current_circuit_index(circuit_index)
+            if self.shots is None and circuit.shots:
+                self._shot_vector = circuit.shots.shot_vector
+                self.shots = circuit.shots.total_shots
```
I think this will mutate the device shots in an unwanted way. In the past I also tried implementing a new measure method for ionq device to avoid using device.shots, but that was quite uneasy and led to a huge PR.
For this update PR, I would suggest we simply locally store the initial `shots` and `shot_vector`, and then after all the usage of shots info done, we restore them, e.g. before `return results`.

splch committed:
Restore device shots after batch_execute to avoid state mutation

Save and restore self.shots and self._shot_vector around the
per-circuit loop so tape-level shots don't leak into device state.

splch commented:
> Finally we have the correct codecov report after having the CI fully pass! Let me scan one last time 🙏 thanks for patience these days

hey thats huge! nice :D

andrijapau commented on CHANGELOG.md:
```diff
@@ -9,6 +9,10 @@
 ### Breaking changes 💔
 
+* Update to v0.4 IonQ API. The job payload now uses `backend` instead of `target`, `type` instead of
+  `format`, and `error_mitigation`/`compilation` are nested under `settings`.
```
Should we link the user to the API reference (https://docs.ionq.com/api-reference/v0.4/introduction)?

JerryChen97 replied:
+1 good idea

andrijapau reviewed (approved):
Changes LGTM, thanks for the effort on getting this in! 💪🏼 🚀

JerryChen97 commented:
Just fast-forwarded again to trigger CI; Will approve as soon as CI green again!

JerryChen97 commented on pennylane_ionq/device.py:
```diff
@@ -221,6 +254,11 @@ def batch_execute(self, circuits):
+        # Use tape-level shots if device shots are not set
+        tape_shots = circuits[0].shots.total_shots if circuits[0].shots else None
+        if self.shots is None and tape_shots is not None:
+            self.job["shots"] = tape_shots
```
```suggestion
            self.job["shots"] = tape_shots  # pylint: disable=access-member-before-definition
```

JerryChen97 commented on pennylane_ionq/device.py:
```diff
@@ -230,9 +268,18 @@ def batch_execute(self, circuits):
+        if self.dry_run:
+            return [[] for _ in circuits]
+
+        original_shots = self.shots
```
```suggestion
        original_shots = self.shots  # pylint: disable=access-member-before-definition
```

JerryChen97 commented on pennylane_ionq/device.py:
```diff
@@ -230,9 +268,18 @@ def batch_execute(self, circuits):
+            return [[] for _ in circuits]
+
+        original_shots = self.shots
+        original_shot_vector = self._shot_vector
```
```suggestion
        original_shot_vector = self._shot_vector  # pylint: disable=access-member-before-definition
```

JerryChen97 reviewed (commented):
Since CI already passed; I feel we can just suppress pylints so that CodeFactor stops false alarming. I tried to suggest some inline ones, but got confused by some of its inline comments :sweat: I probably we can just suppress these warnings at the function level to cover the entire scope.
Then I myself can start a follow-up PR to resolve the issue complained by CF https://github.com/PennyLaneAI/PennyLane-IonQ/issues/172

splch committed:
Add API reference link to changelog, add pylint disable comments

- Link to v0.4 API docs in changelog entry (Andrija + Jerry)
- Add pylint disable=access-member-before-definition on shots access

splch committed:
Merge branch 'master' into new_api_v_04

splch commented:
thank you both for the reviews! i got the latest batch of changes in :)

JerryChen97 reviewed (approved):
🚀 🚀 🚀 🚀
